src/aap_eda/tasks/ruleset.py

```diff
@@ -29,34 +30,125 @@
+            .first()
         )
+        if (
+            not activation
+            or str(activation.status) == ActivationStatus.DELETING.value
+        ):
+            logger.info(f"Activation id: {activation_id} is deleted")
+            return
+
+        if (
+            not activation.is_enabled
+            or str(activation.status) != ActivationStatus.PENDING.value
+        ):
+            logger.info(f"Activation id: {activation_id} cannot be activated")
+            return
+
+        activation.status = ActivationStatus.STARTING
+        activation.save(update_fields=["status"])
+
+        # Note: expect scheduler to set requester as "SCHEDULER"
+        if requester == "SCHEDULER":
+            activation.restart_count += 1
+            activation.save(update_fields=["restart_count", "modified_at"])
+
+    ActivateRulesets().activate(activation)
+
+    activation.refresh_from_db()
+    logger.info(
+        f"Task finished: activation {activation.name}" f" {activation.status}."
+    )
 
 
 @job("default")
-def deactivate(activation_id: int, is_delete: bool = False) -> None:
-    activation = models.Activation.objects.get(id=activation_id)
+def deactivate(
+    activation_id: int, requester: str = "User", delete: bool = False
+) -> None:
+    logger.info(
+        f"Disabling activation id: {activation_id} requested by {requester}"
+    )
+
+    with transaction.atomic():
+        activation = (
+            models.Activation.objects.select_for_update()
+            .filter(id=activation_id)
+            .first()
+        )
+        if not activation:
+            logger.warning(f"Activation {activation_id} is deleted.")
+            return
+
+        if str(activation.status) not in [
+            ActivationStatus.PENDING.value,
+            ActivationStatus.RUNNING.value,
+            ActivationStatus.STARTING.value,
+            ActivationStatus.UNRESPONSIVE.value,
+        ]:
+            logger.warning(
+                f"Cannot deactivate the activation {activation.name} when its "
+                f"status is {activation.status}"
+            )
+            return
+
+        _perform_deactivate(activation)
+
+        if delete:
+            logger.info(f"Activation {activation.name} is deleted")
+            activation.delete()
+        else:
+            activation.current_job_id = None
+            activation.status = ActivationStatus.STOPPED
+            activation.save()
+
+
+@job("default")
+def restart(activation_id: int, requester: str = "User") -> None:
+    logger.info(
+        f"Restarting activation id: {activation_id} requested by {requester}"
+    )
+
+    with transaction.atomic():
+        activation = (
+            models.Activation.objects.select_for_update()
+            .filter(id=activation_id)
+            .first()
+        )
+        if not activation:
+            logger.warning(f"Activation {activation_id} is deleted.")
+            return
+
+        _perform_deactivate(activation)
```

Comment:
pairing with line 93, we can check activation's current status. If it is already stopped/failed etc, we should skip `_perform_deactivate`.